nodes/test/test_ssrf_protection.py

```
@@ -0,0 +1,361 @@
+class TestRedirectSSRFBypass:
+    """Redirect-based SSRF attacks must be blocked.
+
+    An attacker can host a public URL that 302-redirects to a private IP
+    (e.g. the cloud metadata endpoint 169.254.169.254).  The HTTP client
+    must validate each redirect target before following it.
+    """
+
+    def test_redirect_to_private_ip_blocked(self):
+        """A redirect from a public IP to a private IP must be rejected."""
+        with patch('ssrf_protection.socket.getaddrinfo', _fake_getaddrinfo('169.254.169.254')):
+            with pytest.raises(SSRFError, match='private/reserved range'):
```

Ruff (0.15.7)

[warning] Use a single `with` statement with multiple contexts instead of nested `with` statements

(SIM117)